Mathlib/Tactic/StacksAttribute.lean

```
@@ -43,22 +43,59 @@ It extends the `Tag` environment extension with the data `declName, tag, comment
 def addTagEntry {m : Type → Type} [MonadEnv m] (declName : Name) (tag comment : String) : m Unit :=
   modifyEnv (tagExt.addEntry · { declName := declName, tag := tag, comment := comment })
 
-/--
-The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
-This allows `044Q3` and `GH3F6` as possibilities.
--/
-declare_syntax_cat stackTag
+open Parser
+
+/-- `stacksTag` is the node kind of Stacks Project Tags: a sequence of digits and
+uppercase letters. -/
+abbrev stacksTagKind : SyntaxNodeKind := `stacksTag
+
```

```suggestion

/-- Extract the tag if it well-formed -/
def Lean.TSyntax.validateStacksTag (stx : TSyntax stacksTagKind) : Option String :=
  match Syntax.isLit? stacksTagKind stx with
  | some val =>
    if val.all (fun c => c.isDigit || c.isUpper) && val.length == 4 then val
    else none
  | none => none
```

---

You could probably also just throw the error here.

---

```lean
def Lean.TSyntax.validateStacksTag (stx : TSyntax stacksTagKind) : CoreM String := do
  let missing := "Please, enter a Tag after `stacks`."
  let malformed :=
    "Stacks Project tags should be 4 characters and only consist of digits and uppercase letters."
  let some val := Syntax.isLit? stacksTagKind stx | throwError malformed
  if val.length == 0 then throwError missing
  else if val.all (fun c => c.isDigit || c.isUpper) && val.length == 4 then return val
  else throwError malformed
```